pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,423 @@
+
+	// ctx is the ambient context for the replica and is used for remote
+	// traces. It contains the replica/range information. On each trace we
+	// additionally append the unique trace/span IDs.
+	ctx context.Context
+	st  *cluster.Settings
+}
+
+// NewRaftTracer creates a new RaftTracer with the given ambient context for the
+// replica.
+func NewRaftTracer(
+	ctx context.Context, st *cluster.Settings, numRegisteredStore *atomic.Int64,
+) *RaftTracer {
+	return &RaftTracer{ctx: ctx, st: st, numRegisteredStore: numRegisteredStore}
+}
+
+// MaxConcurrentRaftTraces is the maximum number of entries that can be traced
+// at any time on this store. Additional traces will be ignored until the number
+// of traces drops below the limit. Having too many active traces can negatively
+// impact performance as we iterate over all of them for some messages.
+//
+// TODO(baptist): Bump the default to a reasonable value like 10 that balances
+// usefulness with performance impact once we have validated the performance
+// impact.
+var MaxConcurrentRaftTraces = settings.RegisterIntSetting(
+	settings.SystemVisible,
+	"kv.raft.max_concurrent_traces",
+	"the maximum number of tracked raft traces",
+	0,
+	settings.NonNegativeInt,
+)
+
+// reserveSpace checks if should register a new trace. If there are too many
+// registered traces it will not register and return false. The soft invariant
+// is that numRegisteredStore <= numAllowed which can be temporarily violated if
+// the number .  This method will return true if we can add one to the number
+// registered for both the store and replica, otherwise it will return false.
+func (r *RaftTracer) reserveSpace() bool {
+	numAllowed := MaxConcurrentRaftTraces.Get(&r.st.SV)
+	numRegisteredReplica := r.numRegisteredReplica.Load()
+
+	// This can only occur if the numAllowed setting has changed since a
+	// previous call to reserveAll. If this happens flush all our current traces
+	// and don't register this request. Note that when this happens we also
+	// wont't log this request.
+	if numRegisteredReplica > numAllowed {
+		log.Infof(r.ctx, "flushing all traces due to setting change")
+		r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+			r.removeEntry(index)
+			return true
+		})
+		return false
+	}
+
+	if numAllowed == 0 {
+		// This occurs if the setting has changed. If this happens additionally
+		// flush all current traces.
+		return false
+	}
+
+	// The maximum number of traces has been reached for the store. We don't
+	// register tracing and return false.
+	numRegisteredStore := r.numRegisteredStore.Load()
+	if numRegisteredStore >= numAllowed {
+		return false
+	}
+
+	if numRegisteredReplica > numRegisteredStore {
+		panic(fmt.Sprintf("numRegisteredReplica(%d) > numRegisteredStore(%d)", numRegisteredReplica, numRegisteredStore))
+	}
+	if numRegisteredReplica > numAllowed {
+		panic(fmt.Sprintf("numRegisteredReplica(%d) > numAllowed(%d)", numRegisteredReplica, numAllowed))
+	}
```

Comment:
Can you note these invariants in the field definitions? Looks a bit scary though do add these panics. Consider downgrading to `buildutil.CrdbTestBuild`-only checks. Plus, seems not worth increasing the runtime cost.

Reply:
Done